### ISSUE TITLE
Lower or equal comparison inconsistent when local version is present

### ISSUE THREAD

**Issue by lkollar** [pypa/packaging#300]
When the version number contains a local version suffix, the `<=` comparison does not evaluate to True even though the version numbers are identical in the `version` and `SpecifierSet` objects.

Minimal reproducer:
```python
from packaging import version
from packaging.specifiers import SpecifierSet

version.parse('1.0.0+1') in SpecifierSet("==1.0.0") # True
version.parse('1.0.0+1') in SpecifierSet(">=1.0.0") # True
version.parse('1.0.0+1') in SpecifierSet("<=1.0.0") # False
```

I ran into this through `pkg_resources.Requirement` while putting together a resolver based on `resolvelib`, using `pkg_resources` as I started seeing cases where a package with local version identifier (which should satisfy the requirement) ends ends up being ignored.

**Comment by brettcannon** [comment]
The fact that `==` works while `<=` doesn't suggests one of those cases is wrong. Unfortunately I can't tell which is correct from reading https://www.python.org/dev/peps/pep-0440/#local-version-identifiers.

**Comment by di** [comment]
Here's what I found:

### Equality

```python
>>> version.parse('1.0.0+1') in SpecifierSet("==1.0.0")
True
```

This is `True` because it calls `Specifier._compare_equal`, which explicitly ignores the local segment:

https://github.com/pypa/packaging/blob/d2e29ede9ee0e85b66ac550f62766a2a0b0f0fd2/packaging/specifiers.py#L503-L509

so the result is the same as:

```
>>> Version('1.0.0') == Version('1.0.0')`
True
```

### GEQ

```python
version.parse('1.0.0+1') in SpecifierSet(">=1.0.0")
True
```

This is `True` because it calls `Specifier._compare_greather_than_equal`, which does _not_ ignore the local segment:

https://github.com/pypa/packaging/blob/d2e29ede9ee0e85b66ac550f62766a2a0b0f0fd2/packaging/specifiers.py#L521-L524

so the result is the same as:

```python
>>> Version('1.0.0+1') >= Version('1.0.0')
True
```

which is `True` because local versions are considered greater than non-local versions:

```python
>>> Version('1.0.0+1') > Version('1.0.0')
True
```

### LEQ

```python
>>> version.parse('1.0.0+1') in SpecifierSet("<=1.0.0")
False
```

This is `False` because it calls `Specifier._compare_less_than_equal`, which does _not_ ignore the local segment:
 https://github.com/pypa/packaging/blob/d2e29ede9ee0e85b66ac550f62766a2a0b0f0fd2/packaging/specifiers.py#L516-L519

so the result is the same as:

```python
>>> Version('1.0.0+1') <= Version('1.0.0')
False
```

which, again, is `False` because local versions are considered greater than non-local versions:

```python
>>> Version('1.0.0+1') < Version('1.0.0')
False
```

### TL;DR:

I think the question is: should specifier sets ignore local versions when the specifier itself does not have a local version? Right now equality/inequality is ignoring, but the rest are not.

**Comment by lkollar** [comment]
Given that local version numbered releases should be API-compatible with non-local versions, my intuition would be to return True for all the above cases. I think this makes most sense from a user's point of view: if I require public version 1.0.0 of a library, then any such version with an additional local version suffix should satisfy that requirement.

There might be other cases we have to think about here as well. For example, should `Version('1.0.0+1') < Version('1.0.0+2')` respect ordering?

Either way this seems to warrant a clarification in the PEP. What's the best way to decide on the right way forward here? I'm happy to make a PR for the PEP (and the code changes if needed) once there is a decision.

**Comment by di** [comment]
Note that there's a difference here between comparing two `Versions` and testing whether a `Version` is in a given `SpecifierSet`, e.g. `Version('1.0.0') == Version('1.0.0+1')` should always be `False`, while `Version('1.0.0+1') in SpecifierSet("==1.0.0")` should be `True`.

I think we probably want to extend the current behavior of the equality specifier to other specifier comparisons that consider equality as well (LEQ/GEQ).

> There might be other cases we have to think about here as well. For example, should `Version('1.0.0+1') < Version('1.0.0+2')` respect ordering?

Yes, because this is explicitly declared [in the PEP](https://www.python.org/dev/peps/pep-0440/#local-version-identifiers): 

> Comparison and ordering of local versions considers each segment of the local version (divided by a `.`) separately.

and it's already the case:

```
>>> Version('1.0.0+1') < Version('1.0.0+2')
True
>>> Version('1.0.0+1') > Version('1.0.0+2')
False
```

> Either way this seems to warrant a clarification in the PEP.

Looking more closely at the PEP, I found this in https://www.python.org/dev/peps/pep-0440/#version-matching:

> If the specified version identifier is a public version identifier (no local version label), then the local version label of any candidate versions MUST be ignored when matching versions.
>
> If the specified version identifier is a local version identifier, then the local version labels of candidate versions MUST be considered when matching versions, with the public version identifier being matched as described above, and the local version label being checked for equivalence using a strict string equality comparison.

Which is clear enough to me to convince me that this is a bug. Here's my proposed fix: https://github.com/pypa/packaging/pull/304.
